overlord/snapstate/refresh_control.go

```diff
@@ -0,0 +1,173 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/mount"
+	"github.com/snapcore/snapd/overlord/ifacestate/ifacerepo"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+var gateAutoRefreshHookName = "gate-auto-refresh"
+
+func affectedByRefresh(st *state.State, updates []*snap.Info) ([]string, error) {
+	all, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	byBase := make(map[string][]string)
+	for name, snapSt := range all {
+		if !snapSt.Active {
+			delete(all, name)
+			continue
+		}
+		inf, err := snapSt.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		// optimize: do not consider snaps that don't have gate-auto-refresh hook.
+		if inf.Hooks[gateAutoRefreshHookName] == nil {
+			delete(all, name)
+			continue
+		}
+
+		base := inf.Base
+		if inf.Base == "" {
+			base = "core"
+		}
+		byBase[base] = append(byBase[base], inf.InstanceName())
+	}
+
+	affected := make(map[string]bool)
+
+	for _, up := range updates {
+		// add self
+		if all[up.InstanceName()] != nil {
+			affected[up.InstanceName()] = true
+		}
+
+		// snaps that can trigger reboot
+		// XXX: gadget refresh doesn't always require reboot, refine this
+		if up.Type() == snap.TypeKernel || up.Type() == snap.TypeGadget {
+			for _, snapSt := range all {
+				affected[snapSt.InstanceName()] = true
+			}
+			continue
+		}
+		if up.Type() == snap.TypeBase || up.SnapName() == "core" {
+			// affected by refresh of this base snap
+			for _, snapName := range byBase[up.SnapName()] {
+				affected[snapName] = true
+			}
+			continue
+		}
+
+		// no point in further checks
+		if len(affected) == len(all) {
+			break
+		}
+
+		repo := ifacerepo.Get(st)
+
+		// consider slots provided by refreshed snap.
+		for _, slotInfo := range up.Slots {
```

Comment:
I think here we need to special case (ignore?) core and snapd because those will be taken care below?

Reply:
Not addressed this yet, I'm not sure of that.

Reply:
* core can be base which is addressed elsewhere
* on core systems core can be the boot base which is addressed elsewhere,  on core 20 systems snapd can provoke a reboot on its own which I don't think we consider yet

but otherwise I made this comment because as slot providing snaps though, either core or snapd provide the systems slots but those generally give access to unconfined services that are not disrupted by snap updates, unless there is a reboot for other reasons or the mount backend is involved which is covered below